File: algorithms/python/1442_count_triplets_that_can_form_two_arrays_of_equal_xor/count_triplets_that_can_form_two_arrays_of_equal_xor.py
"""
1442. count triplets that can form two arrays of equal xor

https://leetcode.com/problems/count-triplets-that-can-form-two-arrays-of-equal-xor/
https://leetcode-cn.com/problems/count-triplets-that-can-form-two-arrays-of-equal-xor/

形成两个异或相等数组的三元组数目
"""

from typing import List
import collections


def count_triplets(arr: List[int]) -> int:
    """
    哈希表：一重循环
    """
    n = len(arr)
    s = collections.defaultdict(lambda: 0)
    for i, val in enumerate(arr):
        s[i + 1] = s[i] ^ val
    ans = 0
    cnt = collections.defaultdict(lambda: 0)
    total = collections.defaultdict(lambda: 0)
    for k in range(0, n):
        m = cnt.get(s[k + 1], None)
        if m:
            ans += m * k - total[s[k + 1]]
        cnt[s[k]] += 1
        total[s[k]] += k
    return ans

File: algorithms/1442-count-triplets-that-can-form-two-arrays-of-equal-xor/count-triplets-that-can-form-two-arrays-of-equal-xor.go
package problem1442

// countTripletsV1 三重循环
func countTripletsV1(arr []int) (ans int) {
	n := len(arr)
	s := make([]int, n+1)
	for i, val := range arr {
		s[i+1] = s[i] ^ val
	}
	for i := 0; i < n; i++ {
		for j := i + 1; j < n; j++ {
			for k := j; k < n; k++ {
				if s[i] == s[k+1] {
					ans++
				}
			}
		}
	}
	return
}

// countTripletsV2 哈希表：一重循环
func countTripletsV2(arr []int) (ans int) {
	n := len(arr)
	s := make([]int, n+1)
	for i, v := range arr {
		s[i+1] = s[i] ^ v
	}

	cnt := map[int]int{}
	total := map[int]int{}
	for k := 0; k < n; k++ {
		if m, has := cnt[s[k+1]]; has {
			ans += m*k - total[s[k+1]]
		}
		cnt[s[k]]++
		total[s[k]] += k
	}
	return
}
